### PR TITLE
Add preflight checks

### PR DESCRIPTION
**Pull request by ygalblum**
Check that the current kernel is the default one
Check that all devices exist and use ext4 fs

### CODE REVIEW

**Comment by ygalblum** [comment]
@jordigilh PTAL

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/check_device.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+- name: Assert that the filesystem is supported
+  ansible.builtin.assert:
+    that: _mount_info[0].fstype in ['ext4']
+    fail_msg: "Unsupported filesystem '{{ _mount_info[0].fstype }}' on '{{ item.device }}'"
+
+- name: Assert that the filesystem has enough free space
+  ansible.builtin.assert:
+    that: _mount_info[0].block_size * _mount_info[0].block_used < (item.size | ansible.builtin.human_to_bytes)
+    fail_msg: >
+      Requested size {{ item.size }} is smaller than currently used
```
Comment (jordigilh) [inline infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/check_device.yaml:19]: suggestion: `item.size | ansible.builtin.human_to_bytes`. The point is to show values of the same unit, and here you might be referencing Gb to bytes.

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/check_kernel.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+- name: Get default kernel
+  ansible.builtin.command:
+    cmd: /sbin/grubby --default-kernel
```
Comment (jordigilh) [inline infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/check_kernel.yaml:3]: Do you expect the default kernel to be different than the one currently running?

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/check_kernel.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  register: _grubby_rc
+  changed_when: false
+
+- name: Parse default kernel version
+  ansible.builtin.set_fact:
+    _default_kernel: "{{ ((((_grubby_rc.stdout_lines[0] | split('/'))[2] | split('-'))[1:]) | join('-')) | trim }}"
+
+- name: Check the values
+  ansible.builtin.assert:
+    that: _default_kernel == kernel_version
```
Comment (jordigilh) [inline infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/check_kernel.yaml:13]: Oh I see... Nevermind my comment 😄

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/check_device.yaml (hunk 1 above)
Comment (ygalblum) [reply infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/check_device.yaml:19]: Are you referring to the error message? If yes, then this is the value provided by the user and therefore the one I think should be in the message.

Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/check_device.yaml:19]: Ah I see. Sorry I got confused with `human_readable` with `human_to_bytes`. Nothing to see here.. move along 😄

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/check_device.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+- name: Get the mount point info
+  ansible.builtin.set_fact:
+    _mount_info: "{{ ansible_facts.mounts | selectattr('device', 'equalto', item.device) }}"
```
Comment (jordigilh) [inline infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/check_device.yaml:3]: Does it matter if the device has a mount point or that it's mounted when the facts are retrieved? I think this check should only care that the device name exists in the host. WDYT?

Comment (jordigilh) [reply infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/check_device.yaml:3]: Also, an LV can be found as `/dev/mapper/<vg_name>-<lv_name>` as well as `/dev/<vg_name>/<lv_name>`. This check would render the later one as not valid.

**Comment by jordigilh** [comment]
LGTM. Minor suggestion to give more flexibility in passing the device name based on `/dev/<vg_name>/<lv_name>` as well as `/dev/mapper/<vg_name>-<lv_name>`, since both are valid.

File: infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/check_device.yaml (hunk 4 above)
Comment (ygalblum) [reply infra.lvm_tools/shrink_lv/roles/shrink_lv/tasks/check_device.yaml:3]: You're right about the fact that it's enough that is exists. But, we use the `mount_info` for additional verifications such as the filesystem type and bytes used. So, if we don't get the path that is actually mounted, we can't check that.

**Comment by swapdisk** [comment]
@ygalblum, I tried running this with the following vars:
```
shrink_lv_devices:
  - device: /dev/mapper/VolGroup00-homeVol
    size: 1.8G
```
Observed the following output on the console while rebooting:
```
         Starting dracut pre-mount hook...
[  OK  ] Started File System Check on /dev/mapper/VolGroup00-rootVol.
[    7.786091] dracut-pre-mount[378]: Size is not a multiple of 512. Try using 1932734976 or 1932735488.
[    7.810192] dracut-pre-mount[378]: Invalid argument for --size: 1932735284b
[  OK  ] Started dracut pre-mount hook.
[    7.812820] dracut-pre-mount[378]: Error during parsing of command [   12.859521] EXT4-fs (dm-0): mounted filesystem with ordered data mode. Opts: (null)
line.
         Mounting /sysroot...
```
Boot continued successfully, but the filesystem and logical volume were not shrunk. 

Tried again with `size: 1932734976` and it worked as expected.

**Comment by ygalblum** [comment]
@swapdisk thanks for this.
Since the issue is not related to this PR, I opened a separate one: https://github.com/ygalblum/ansible-bash-shrink-lv/pull/5
Note that I aligned the size to 4K in order to be safer.

**Comment by heatmiser** [comment]
Tested in AAP2 via project requirements.yml...

`- name: https://github.com/ygalblum/ansible-bash-shrink-lv#infra.lvm_tools/shrink_lv
  type: git
  version: prelight-checks`

...with playbook calling role as:

`  roles:
    - name: infra.shrink_lv.shrink_lv`

and job template vars as:

shrink_lv_devices:
  - device: /dev/mapper/VolGroup00-homeVol
    size: 1G

2G logical volume VolGroup00-homeVol was reduced to 1G

Please merge.

**Pull request merged**